File: app/src/main/java/org/senatov/model/tree/DirTreeModel.java
/*
 * DirTreeModel.java — manages tree state for dir compare panel.
 * Builds flat list from tree for ListView display.
 * Handles expand/collapse by relativePath.
 * Iakov Senatov, 2026
 */
package org.senatov.model.tree;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.senatov.model.CompareLineItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


@Slf4j
public final class DirTreeModel {

    @Getter
    private final List<DirTreeNode> roots;
    private final Set<String> expandedPaths = new HashSet<>();

    public DirTreeModel(List<DirTreeNode> roots) {
        this.roots = roots;
    }


    public void toggleExpand(String relativePath) {
        if (expandedPaths.contains(relativePath)) {
            expandedPaths.remove(relativePath);
            log.debug("collapsed: {}", relativePath);
        } else {
            expandedPaths.add(relativePath);
            log.debug("expanded: {}", relativePath);
        }
    }


    public void expandAll() {
        expandAllRecursive(roots);
        log.info("expanded all dirs, count={}", expandedPaths.size());
    }


    public void collapseAll() {
        expandedPaths.clear();
        log.info("collapsed all dirs");
    }


    public boolean isExpanded(String relativePath) {
        return expandedPaths.contains(relativePath);
    }


    public List<CompareLineItem> toFlatList() {
        List<CompareLineItem> result = new ArrayList<>();
        for (DirTreeNode root : roots) {
            flattenNode(root, result);
        }
        return result;
    }


    private void flattenNode(DirTreeNode node, List<CompareLineItem> result) {
        boolean expanded = node.isDirectory() && expandedPaths.contains(node.getRelativePath());

        CompareLineItem item = new CompareLineItem(
                0, node.getName(), node.getStatus(),
                node.getDepth(), node.isDirectory(), node.getRelativePath(),
                node.getSize(), node.getLastModifiedMs()
        );
        item.setExpanded(expanded);
        result.add(item);

        if (!expanded || !node.isDirectory()) {
            return;
        }

        List<DirTreeNode> sorted = new ArrayList<>(node.getChildren());
        Collections.sort(sorted);

        for (DirTreeNode child : sorted) {
            flattenNode(child, result);
        }
    }


    private void expandAllRecursive(List<DirTreeNode> nodes) {
        for (DirTreeNode node : nodes) {
            if (node.isDirectory()) {
                expandedPaths.add(node.getRelativePath());
                expandAllRecursive(node.getChildren());
            }
        }
    }
}

File: app/src/main/kotlin/org/senatov/model/tree/DirTreeModel.kt
/*
 * DirTreeModel — manages tree state 4 dir compare panel.
 * flat list from tree for ListView display.
 * expand/collapse by relativePath.
 * Iakov Senatov, 2026
 */
package org.senatov.model.tree

import org.senatov.helpers.log.LogTag
import org.senatov.model.CompareLineItem
import org.senatov.model.LineContent
import org.senatov.model.TreeDisplayState
import org.slf4j.LoggerFactory


class DirTreeModel(val roots: List<DirTreeNode>) {

    private val log = LoggerFactory.getLogger(DirTreeModel::class.java)
    private val expandedPaths = mutableSetOf<String>()


    fun toggleExpand(relativePath: String) {
        log.debug(LogTag.UI, "toggleExpand(relativePath={})", relativePath)
        if (relativePath in expandedPaths) {
            expandedPaths.remove(relativePath)
            log.debug(LogTag.UI, "collapsed {}", relativePath)
        } else {
            expandedPaths.add(relativePath)
            log.debug(LogTag.UI, "expanded {}", relativePath)
        }
    }


    fun expandAll() {
        log.debug(LogTag.UI, "expandAll()")
        expandAllRecursive(roots)
        log.info(LogTag.UI, "expanded all count={}", expandedPaths.size)
    }


    fun collapseAll() {
        log.debug(LogTag.UI, "collapseAll()")
        expandedPaths.clear()
        log.info(LogTag.UI, "collapsed all")
    }


    fun isExpanded(relativePath: String): Boolean {
        log.debug(LogTag.UI, "isExpanded(relativePath={})", relativePath)
        return relativePath in expandedPaths
    }


    fun toFlatList(): List<CompareLineItem> {
        log.debug(LogTag.UI, "toFlatList()")
        val result = mutableListOf<CompareLineItem>()
        for (root in roots) {
            flattenNode(root, result)
        }
        return result
    }


    private fun flattenNode(node: DirTreeNode, result: MutableList<CompareLineItem>) {
        log.debug(
            LogTag.UI,
            "flattenNode(relativePath={}, resultSize={})",
            node.relativePath,
            result.size
        )
        val expanded = node.isDirectory && node.relativePath in expandedPaths
        val item = CompareLineItem(
            content = LineContent(number = 0, text = node.name),
            status = node.status,
            treeState = TreeDisplayState(details = node.details, isExpanded = expanded),
        )
        result.add(item)
        if (!expanded || !node.isDirectory) return
        for (child in node.children.sorted()) {
            flattenNode(child, result)
        }
    }


    private fun expandAllRecursive(nodes: List<DirTreeNode>) {
        log.debug(LogTag.UI, "expandAllRecursive(nodes={})", nodes.size)
        for (node in nodes) {
            if (node.isDirectory) {
                expandedPaths.add(node.relativePath)
                expandAllRecursive(node.children)
            }
        }
    }
}
